Implement mana/stamina system

lowkangn opened the pull request:
Resolves #43 

Lets:
* Add resource bars to the in-game UI to represent the Knight's stamina level and Dragon's mana level.
* Add a ConsumableResource script to all actors. This script encapsulates the resource the actor uses, and the resource amount can be set in the Inspector.
* Add a resource cost to combat abilities. The cost for an ability can be set in the Inspector. An actor cannot execute an ability if they do not have enough resource to pay for the cost. If an ability should not cost any resource, the cost can be set to 0. (Particularly for enemies)
* Implement visual updates for the stamina and mana bars to reflect the players' stamina and mana levels respectively. When a player uses an ability, the resource will have a short delay before it will start regenerating. Executing another ability again will interrupt the regeneration.
* Change the timers in Buff and EscapeEvent to be coroutines instead.
* Disable mana regeneration for the Dragon when mounted.
* Set resource level to zero and disable resource regeneration when the player dies.
* Restart resource regeneration immediately when the player is revived. The resource level will start from zero.
* Add a stamina cost for blocking an attack as the Knight.

Things to consider: (in future PRs)
* ~~Disabling the mana regeneration for the Dragon when it is mounted. This will further balance the Dragon riding mechanic which may be too overpowered.~~
* Including resource costs for certain enemy abilities. Currently, all enemies do not have any resource pools and all their abilities have 0 cost.

lowkangn committed:
Change timers in Buff and Escape Menu to use coroutines

Let's:
* Change the timers implemented in Buff and EscapeMenu scripts to use coroutines instead. This will remove the unnecessary checks made in Update method.

lowkangn committed:
Add UI for stamina/mana bars

Let's:
* Add sliders to the in-game UI to show the stamina and mana levels of the Knight and Dragon respectively.
* Add a new script ConsumableResource for simulating the mana/stamina system. (Not fully implemented)

lowkangn committed:
Merge branch 'master' into branch-mana-and-stamina

lowkangn committed:
Implement stamina/mana system

Lets:
* Add a ConsumableResource script to all actors. This script encapsulates the resource the actor uses, and the resource amount can be set in the Inspector.
* Add a resource cost to combat abilities. The cost for an ability can be set in the Inspector. An actor cannot execute an ability if they do not have enough resource to pay for the cost. If an ability should not cost any resource, the cost can be set to 0. (Particularly for enemies)
* Implement visual updates for the stamina and mana bars to reflect the players' stamina and mana levels respectively. When a player uses an ability, the resource will have a short delay before it will start regenerating. Executing another ability again will interrupt the regeneration.

Things to consider: (in future PRs)
* Disabling the mana regeneration for the Dragon when it is mounted. This will further balance the Dragon riding mechanic which may be too overpowered.
* Including resource costs for certain enemy abilities. Currently, all enemies do not have any resource pools and all their abilities have 0 cost.

kenneth-fung commented on Assets/Scripts/Actor Components/Combat/Combat Abilities/MeleeAttackAbility.cs:
```diff
@@ -17,8 +17,9 @@ public override void Execute(Combat combat, params object[] parameters)
             combat.CombatStateMachine.ChangeState(new ReadyAttackState(combat, readyDuration, ReadyCallback));
         }
-        else
+        else if (combat.Resource.CanConsume(resourceCost))
         {
+            combat.Resource.Consume(resourceCost);
             combat.CombatStateMachine.ChangeState(new MeleeAttackState(combat, attackEffectArea));
```
Some actors (enemies) are required to ready their melee attack before executing it. The actual execution occurs in the `ReadyCallback` method below this `Execute` method, so the resource check needs to occur there as well. Since there would be repeated logic in `Execute` and `ReadyCallback`, you can also consider consolidating these lines into a private method and call it from here and `ReadyCallback`.

kenneth-fung commented on Assets/Scripts/Actor Components/Combat/Combat Abilities/RangedAttackAbility.cs:
```diff
@@ -36,8 +36,10 @@ public override void Execute(Combat combat, params object[] parameters)
                 combat, target, projectilePathDisplay, timeToLock, readyDuration, ReadyCallback));
         }
-        else
+        else if (combat.Resource.CanConsume(resourceCost))
         {
+            combat.Resource.Consume(resourceCost);
+
             Vector2 direction = (Vector2)parameters[0];
             combat.CombatStateMachine.ChangeState(new RangedAttackState(
```
The same issue with the `ReadyCallback` as in `MeleeAttackAbility` exists here.

kenneth-fung commented on Assets/Scripts/Actor Components/ConsumableResource.cs:
```diff
@@ -0,0 +1,66 @@
+        yield return new WaitForSeconds(delayBeforeRegen);
+
+        regenerateResourceEvent.Invoke();
+        while (CurrentAmount < maxAmount)
+        {
+            CurrentAmount = Mathf.Min(maxAmount, CurrentAmount + maxAmount / 50);
+            yield return regenTick;
+        }
```
It might be better to use `Time.deltaTime` and `yield return null` to tick the regeneration. This would allow us to define and set a specific regeneration speed as a serialized field in the inspector and use it here instead of relying on magic numbers. It would also help to smoothen the rate of increase.

kenneth-fung commented on Assets/Scripts/Actor Components/ConsumableResource.cs:
```diff
@@ -0,0 +1,66 @@
+    [SerializeField] private ConsumeResourceEvent consumeResourceEvent;
+    [SerializeField] private PhotonView photonView;
+
+    private readonly float delayBeforeRegen = 1.5f;
```
Consider making this a serialized field so it can be adjusted in the inspector.

kenneth-fung commented on Assets/Scripts/UI/ResourceBar.cs:
```diff
@@ -0,0 +1,43 @@
+
+    public IEnumerator Regenerate()
+    {
+        while (slider.value < maxValue)
+        {
+            slider.value = Mathf.Min(maxValue, slider.value + maxValue / 50);
+            yield return regenTick;
+        }
```
Same issue here as with the coroutine in `ConsumableResource`.

kenneth-fung reviewed (changes requested):
Overall methodology looks good and it tests well. However, some under-the-hood changes requested.

An issue not mentioned below is that resource regeneration continues while an actor is dead. Ideally, upon entering `DeathState` the actor's consumable resource should drop to zero and have its regeneration disabled until it exits `ReviveState`. That logic could also be used for the dragon when mounting. But I'm alright with merging without resolving these two issues first, they can come in a future PR no problem.

lowkangn committed:
Change resource regeneration to use deltaTime

Let's:
* Use deltaTime instead of WaitForSeconds for a smoother rate of resource regeneration.
* Add serializable fields for regeneration delay and regeneration speed.

lowkangn committed:
Add a stamina cost for blocking an attack

Let's:
* Add a stamina cost to the Knight's block. The Knight is still able to block even with 0 stamina. However, every attack that is blocked will incur a stamina cost.

lowkangn committed:
Fix issues with resource being depleted incorrectly

Players can deplete their resources multiple times in quick succession when clicking on action inputs too quickly, even though the actions are not actually being carried out on screen.

Let's:
* Shift the consumption of resource into MeleeAttackState and RangedAttackState. The resource will now only be consumed if the ability is properly executed.
* Change the order of the check for resource level in MeleeAttackAbility and RangedAttackAbility. The ability will only be executed if there is enough resource to pay for the cost.
* Change the resource consumption for blocking an attack to make use of blockHitEvent in BlockHitState.

lowkangn committed:
Add resource handling for different combat states

Let's:
* Set resource level to zero when the player dies. Resource regeneration is also turned off.
* Restart resource regeneration immediately when the player is revived. The player's resource is not replenished when revived to create some sense of vulnerability.
* Turn off mana regeneration for the Dragon when it is mounted. When the Knight dismounts, there is the usual delay before mana will start regenerating.

kenneth-fung commented on Assets/Scripts/Input/DragonPlayerController.cs:
```diff
@@ -50,6 +52,9 @@ protected override void OnEnable()
         if (photonView.IsMine)
         {
             Combat.Health.UpdateHealthEvent.AddListener(healthUI.UpdateDragonHealthUI);
+            Combat.Resource.UpdateResourceEvent.AddListener(manaUI.UpdateManaUI);
+            Combat.Resource.RegenerateResourceEvent.AddListener(manaUI.RegenerateManaUI);
+            Combat.Resource.StopRegenResourceEvent.AddListener(manaUI.StopRegenManaUI);
```
When adding listeners to an event in `OnEnable`, it's advisable to remove them in `OnDisable` as it can result in a memory leak otherwise.

lowkangn replied:
👍

kenneth-fung commented on Assets/Scripts/Input/KnightPlayerController.cs:
```diff
@@ -49,6 +51,9 @@ protected override void OnEnable()
         if (photonView.IsMine)
         {
             Combat.Health.UpdateHealthEvent.AddListener(healthUI.UpdateKnightHealthUI);
+            Combat.Resource.UpdateResourceEvent.AddListener(staminaUI.UpdateStaminaUI);
+            Combat.Resource.RegenerateResourceEvent.AddListener(staminaUI.RegenerateStaminaUI);
+            Combat.Resource.StopRegenResourceEvent.AddListener(staminaUI.StopRegenManaUI);
```
Same issue here with removing the listeners in `OnDisable`.

kenneth-fung reviewed (changes requested):
Looks good, just a small change requested.

lowkangn committed:
Apply suggestions from code review

kenneth-fung reviewed (approved):
Looks good.
